File: simple_loader.cpp
#include <iostream>
#include <thread>
#include <chrono>

void loadingBar(int totalSteps, const std::string& prefix = "Loading", const std::string& suffix = "Complete", int length = 50, char fill = '█') {
    // Function to print the loading bar
    for (int step = 0; step <= totalSteps; ++step) {
        float percent = (float(step) / totalSteps) * 100;
        int filledLength = int(length * step / totalSteps);
        std::string bar(filledLength, fill);
        bar.append(length - filledLength, '-');
        std::cout << "\r" << prefix << " |" << bar << "| " << percent << "% " << suffix << std::flush;
        std::this_thread::sleep_for(std::chrono::milliseconds(100)); // Simulate work
    }
    std::cout << std::endl; // New line after completion
}

void spinningLoader(int duration, const std::string& message = "Processing", const std::string& spinnerChars = "|/-\\") {
    // Function to display the spinning loader
    bool stopSpinner = false;
    int idx = 0;
    std::thread spinner([&](){
        while (!stopSpinner) {
            std::cout << "\r" << message << " " << spinnerChars[idx] << std::flush;
            idx = (idx + 1) % spinnerChars.length();
            std::this_thread::sleep_for(std::chrono::milliseconds(100)); // Spinner speed
        }
    });

    // Simulate work duration
    std::this_thread::sleep_for(std::chrono::seconds(duration));
    stopSpinner = true;
    spinner.join();
    std::cout << "\rDone!       "; // Clear the line after completion
}

int main() {
    // Example usage of loading bar
    loadingBar(100, "Downloading", "Complete", 50);

    // Example usage of spinning loader
    spinningLoader(5, "Processing", "|/-\\");
}

File: SimpleLoader.js
class SimpleLoader {
    /**
     * A simple yet effective loading bar and spinner utility.| By https://github.com/JMitander/ | (MIT License) 
     */

    static loadingBar(totalSteps, prefix = 'Loading', suffix = 'Complete', length = 50, fill = '█') {
        /**
         * Displays a progress bar.
         *
         * @param {number} totalSteps - Total steps for completion.
         * @param {string} prefix - Prefix string.
         * @param {string} suffix - Suffix string.
         * @param {number} length - Length of the bar.
         * @param {string} fill - Bar fill character.
         */
        const printBar = (step) => {
            const percent = (step / totalSteps) * 100;
            const filledLength = Math.floor(length * step / totalSteps);
            const bar = fill.repeat(filledLength) + '-'.repeat(length - filledLength);
            process.stdout.write(`\r${prefix} |${bar}| ${percent.toFixed(1)}% ${suffix}`);
        };

        for (let step = 0; step <= totalSteps; step++) {
            printBar(step);
            setTimeout(() => {}, 100); // Simulates work with a delay
        }
        console.log(); // New line after completion
    }

    static spinningLoader(duration = 5, message = 'Processing', spinnerChars = '|/-\\') {
        /**
         * Displays a spinning loader.
         *
         * @param {number} duration - Time duration in seconds.
         * @param {string} message - Message to display alongside the spinner.
         * @param {string} spinnerChars - Characters to use for the spinner.
         */
        let stopSpinner = false;
        const spinner = () => {
            let idx = 0;
            const interval = setInterval(() => {
                if (stopSpinner) {
                    clearInterval(interval);
                    process.stdout.write('\rDone!       ');
                    return;
                }
                process.stdout.write(`\r${message} ${spinnerChars[idx]}`);
                idx = (idx + 1) % spinnerChars.length;
            }, 100);
        };

        spinner();
        setTimeout(() => {
            stopSpinner = true;
        }, duration * 1000); // Duration in milliseconds
    }
}

// Loading bar example (remove this in your actual script)
SimpleLoader.loadingBar(100, 'Downloading', 'Complete', 50);

// Example usage (remove this in your actual script)
SimpleLoader.spinningLoader(5, 'Processing', '|/-\\');
